cmd/rofl/build/artifacts.go

```diff
@@ -0,0 +1,304 @@
+			if err = os.MkdirAll(path, header.FileInfo().Mode()); err != nil {
+				return fmt.Errorf("failed to create directory: %w", err)
+			}
+		case tar.TypeLink:
+			// Hard link.
+			var linkPath string
+			linkPath, err = cleanupPath(header.Linkname)
+			if err != nil {
+				return err
+			}
+
+			if err = os.Link(linkPath, path); err != nil {
+				return fmt.Errorf("failed to create hard link: %w", err)
+			}
+		case tar.TypeSymlink:
+			// Symbolic link.
+			if err = os.Symlink(header.Linkname, path); err != nil {
+				return fmt.Errorf("failed to create soft link: %w", err)
+			}
+		case tar.TypeChar, tar.TypeBlock, tar.TypeFifo:
+			// Device or FIFO node.
+			if err = extractHandleSpecialNode(path, header); err != nil {
+				return err
+			}
+		case tar.TypeReg:
+			// Regular file.
+			if err = os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
+				return fmt.Errorf("failed to create parent directory: %w", err)
+			}
+
+			var fh *os.File
+			fh, err = os.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_TRUNC, header.FileInfo().Mode())
+			if err != nil {
+				return fmt.Errorf("failed to create file: %w", err)
+			}
+			if _, err = io.Copy(fh, rd); err != nil { //nolint:gosec
+				fh.Close()
+				return fmt.Errorf("failed to copy data: %w", err)
+			}
+			fh.Close()
+		default:
+			// Skip unsupported types.
+			continue
+		}
+	}
+
+	// Update all modification times at the end to ensure they are correct.
+	for path, mtime := range modTimes {
+		if err = extractChtimes(path, mtime, mtime); err != nil {
+			return fmt.Errorf("failed to change file '%s' timestamps: %w", path, err)
+		}
+	}
+
+	return nil
+}
+
+// copyFile copies the file at path src to a file at path dst using the given mode.
+func copyFile(src, dst string, mode os.FileMode) error {
+	sf, err := os.Open(src)
+	if err != nil {
+		return fmt.Errorf("failed to open '%s': %w", src, err)
+	}
+	defer sf.Close()
+
+	df, err := os.OpenFile(dst, os.O_RDWR|os.O_CREATE|os.O_TRUNC, mode)
+	if err != nil {
+		return fmt.Errorf("failed to create '%s': %w", dst, err)
+	}
+	defer df.Close()
+
+	_, err = io.Copy(df, sf)
+	return err
+}
+
+// computeDirSize computes the size of the given directory.
+func computeDirSize(path string) (int64, error) {
+	var size int64
+	err := filepath.WalkDir(path, func(_ string, d fs.DirEntry, derr error) error {
+		if derr != nil {
+			return derr
+		}
+		fi, err := d.Info()
+		if err != nil {
+			return err
+		}
+		size += fi.Size()
+		return nil
+	})
+	if err != nil {
+		return 0, err
+	}
+	return size, nil
+}
+
+// createExt4Fs creates an ext4 filesystem in the given file using directory dir to populate it.
+//
+// Returns the size of the created filesystem image in bytes.
+func createExt4Fs(fn, dir string) (int64, error) {
+	// Compute filesystem size in bytes.
+	fsSize, err := computeDirSize(dir)
+	if err != nil {
+		return 0, err
+	}
+	fsSize /= 1024                // Convert to kilobytes.
+	fsSize = (fsSize * 150) / 100 // Scale by overhead factor of 1.5.
```

Comment:
What is the typical size of the generated filesystems that you've tried so far?

Also, might be a good idea to add a check to make sure that `fn` is not an existing block device, so we don't accidentally trash a user's drive :)
(Currently, in the only place this function is called, the filename is hardcoded to `rootfs.ext4`, so it's not a problem, but it might be a problem in the future if we maybe have configurable filenames :man_shrugging:)